meson.build

```diff
@@ -89,7 +89,7 @@ if get_option('post_install')
     meson.add_install_script(
       env_util.path(),
       'GSCHEMADIR=' + gschemadir,
-      join_paths(meson.source_root(), 'build-aux', 'meson', 'post-install.sh')
+      join_paths(meson.project_source_root(), 'build-aux', 'meson', 'post-install.sh')
```

Comment:
If you bump just a bit further to 0.57.0 you can drop this entirely, and use https://mesonbuild.com/Gnome-module.html#gnomepost_install

Reply:
Ah that's cool, I didn't know about that function!

I think we need the option to set the GSchema directory [for NixOS](https://github.com/NixOS/nixpkgs/blob/18a07645e5428709a48e617fd52dc8aedcc85e4e/pkgs/desktops/gnome/extensions/gsconnect/default.nix#L63), unless there's a different way to do that in NixOS.

Reply:
I don't really know how Nix works, but what are they doing for the entire rest of the gome ecosystem, which do in fact install to the *only* gnome-approved location (overridable via XDG_DATA_DIRS, not by specifying random paths)?

Reply:
The Nix manual shows this, which says it sets things up via XDG_DATA_DIRS for you: https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-hooks

Reply:
The Nix manual _also_ shows [this](https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-common-issues), though:

> #### 17.12.4.4. I am packaging a project that cannot be wrapped, like a library or GNOME Shell extension.
> You can rely on applications depending on the library setting the necessary environment variables but that is often easy to miss. Instead we recommend to patch the paths in the source code whenever possible. Here are some examples:
> 
> * [Replacing a GI_TYPELIB_PATH in GNOME Shell extension](https://github.com/NixOS/nixpkgs/blob/7bb8f05f12ca3cff9da72b56caa2f7472d5732bc/pkgs/desktops/gnome-3/core/gnome-shell-extensions/default.nix#L21-L24) – we are using substituteAll to include the path to a typelib into a patch.
> 
> * The following examples are hardcoding GSettings schema paths. To get the schema paths we use the functions
> 
>     * glib.getSchemaPath Takes a nix package attribute as an argument.
> 
>     * glib.makeSchemaPath Takes a package output like $out and a derivation name. You should use this if the schemas you need to hardcode are in the same derivation.
> 
>     [Hard-coding GSettings schema path in Vala plug-in (dynamically loaded library)](https://github.com/NixOS/nixpkgs/blob/7bb8f05f12ca3cff9da72b56caa2f7472d5732bc/pkgs/desktops/pantheon/apps/elementary-files/default.nix#L78-L86) – here, substituteAll cannot be used since the schema comes from the same package preventing us from pass its path to the function, probably due to a [Nix bug](https://github.com/NixOS/nix/issues/1846).
> 
>     [Hard-coding GSettings schema path in C library](https://github.com/NixOS/nixpkgs/blob/29c120c065d03b000224872251bed93932d42412/pkgs/development/libraries/glib-networking/default.nix#L31-L34) – nothing special other than using [Coccinelle patch](https://github.com/NixOS/nixpkgs/pull/67957#issuecomment-527717467) to generate the patch itself.

Reply:
(Configuring `XDG_DATA_DIRS` per-package only works for installed applications, it can't be set per-Extension because they all get launched in the GNOME Shell process environment.)

Reply:
Yeah, unfortunately the Meson’s `gnome.post_install` will not work for us. Ideally it would accept `gsettings_schema_dirs` option, added that to my TODO list.

But I recognize that relying on projects to compile the schemas is a bit of a hack – the only reason we have not gotten to doing it ourselves is that relying on the project’s build system works pretty well in practice – so I would not mind if GSconnect switched to `gnome.post_install`. We can always compile the schemas in the custom location ourselves for the packages that cannot do that and it is not a huge deal. The important thing is there is a single place where we can replace the path where GSconnect looks for the schemas so that it can find it after we move them – be it an option, or a constant definition in the source code we can patch.

Reply:
Quite frankly I don't understand why absolutely any of this is necessary. Doesn't Nix already have dedicated per-package datadirs for exactly this reason? If building XDG_DATA_DIRS based on that isn't good enough, then considering this extension already hardcodes a gschema from_file (??) why isn't configuring it based on the datadir sufficient?

Why does there need to be another option in addition to the datadir?

Reply:
I *think* the only two places this matters is [`config.js.in`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/data/config.js.in) and [`metadata.json.in`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/data/metadata.json.in) where we replace the `@GSETTINGS_SCHEMA_DIR@` placeholder, which should only come from the variable `gschemadir` in `/meson.build`.

IIRC `-Dgsettings_schemadir` was only for NixOS, so if we remove that we could just have a line like:

```meson
gschemadir = join_paths(datadir, 'glib-2.0', 'schemas')
```

I guess that would be the easiest to patch?

Reply:
> Nix has a mode of operation where it creates so called “environment” by combining the trees of multiple packages into a single tree using symlinks. If we used the standard GSettings schema path, `gschemas.compiled` files from different packages would conflict.

I don't understand how this is related.

You have multiple prefixes, multiple datadirs, and multiple gschema directories. The trees of multiple packages will automatically conflict on gschemas.compiled, whether that directory is called:
- `{hash-of-gsconnect}/share/glib-2.0/schemas/` or
- `{hash-of-gsconnect}/share/gsettings-schemas/gsconnect/glib-2.0/schemas`

Elsewhere, you build a list of XDG_DATA_DIRS to collect various directories as inputs to a program, and create a wrapper program that sets the variable before running the program. Presumably, this would also be how you handle
> so called "environment[s]"

But failing that, assuming that it *is* hardcoded... why is it not sufficient to hardcode `{hash-of-gsconnect}/share/glib-2.0/schemas/` inside the program, such that the schema is loaded directly from there without looking anywhere other than inside the prefix in question?

Reply:
But yes, if you have a mode where you allow overlaying directory trees from multiple applications using a shared cache file, you should be taking that into account. What do you do for mime packages? *.desktop files?

Reply:
> I _think_ the only two places this matters is [`config.js.in`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/data/config.js.in) and [`metadata.json.in`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/data/metadata.json.in) where we replace the `@GSETTINGS_SCHEMA_DIR@` placeholder, which should only come from the variable `gschemadir` in `/meson.build`.
> 
> IIRC `-Dgsettings_schemadir` was only for NixOS, so if we remove that we could just have a line like:
> 
> ```meson
> gschemadir = join_paths(datadir, 'glib-2.0', 'schemas')
> ```
> 
> I guess that would be the easiest to patch?

Okay, this has been pushed but is waiting on GSConnect/gsconnect-ci#9

Reply:
> You have multiple prefixes, multiple datadirs, and multiple gschema directories. The trees of multiple packages will automatically conflict on gschemas.compiled, whether that directory is called:

Not necessarily. The path is different for each package – so unless one tries to have multiple variants of the same package in the environment, there won’t be conflicts:

```
/nix/store/x6qf6457yxbcjbmgriyzy06clc39dh6a-geary-40.0/share/gsettings-schemas/geary-40.0/glib-2.0/schemas/gschemas.compiled
/nix/store/m2d0yjh7xmv2ccbadqhv79gv4ckr5isy-evince-42.2/share/gsettings-schemas/evince-42.2/glib-2.0/schemas/gschemas.compiled
```

> But yes, if you have a mode where you allow overlaying directory trees from multiple applications using a shared cache file, you should be taking that into account. What do you do for mime packages? *.desktop files?

Caches are generally not necessary for programs to work so packages do not include them and we have hooks that will remove them if they are generated. Instead, NixOS is responsible for generating them in the system environment – Nix’s and Nixpkgs’ `buildEnv` functions are too low level for that.

Reply:
And this is still completely ignoring the fact that I suggested, you do not need to set up data dirs at all. If the software ***hardcodes*** `/nix/store/x6qf6457yxbcjbmgriyzy06clc39dh6a-geary-40.0/share/glib-2.0/schemas/` then it will just point there regardless of whether you set up an environment merger and regardless of the value of XDG_DATA_DIRS. So I really don't understand the problem that necessitates adding new options to all software.

Reply:
> That's also why Meson doesn't run them if DESTDIR is set :) because no other distro runs them in the build environment either.

Right, I am aware. We do not usually set `DESTDIR` because we just have no need for it – the package build can only write to the Nix store paths it owns. The side effect of projects compiling their schemas was a just nice bonus, allowing us to not deal with the issue for the time being.

> But then you're not overlaying them together! You have to manually set XDG_DATA_DIRS to use these totally third-party (as far as glib is concerned) schema directories.

The individual packages usually already know the path to their datadir (either it is set by a wrapper or hardcoded using a patch). But the overlaying is **not** intended for the individual packages (for the most part, though sometimes, the mechanism is used for packages that use plug-ins, then there might need to be extra wrappers) – the purpose is to have single entry point we can symlink to a global path (e.g. `/run/current-system/sw/` on NixOS) whose subdirectories can be added to system-wide variables like `PATH` or `XDG_DATA_DIRS`. The symlink then can be swapped without having change the environment variables and restart everything.

> You cannot have it both ways. Either the goal is to merge the tree and fake out glib by telling glib that "yes, /glib-2.0/schemas/ is right here, it has all the files", and multiple packages conflict on the gschemas, or, you rely on XDG_DATA_DIRS to tell glib where to find schemas, in which case you do not need to merge the tree at all, just set the environment variable pointing at the store versions

Since the merged trees is not used to find the schemas, it is irrelevant for this.

> and exclude that directory from the merger (or map it specially).

That would be preferred but, as I mentioned, the component responsible for merging the trees is very low-level and does not care about tree semantics. Ideally we would introduce content-aware merge function but since we do not have it yet, we are stuck with this ancient hack. I have it on my TODO list but the separation of concerns between Nix, Nixpkgs and NixOS makes it somewhat difficult to decide what project should take care of that.

> You are creating extra work for yourself, and extra work for downstream developers, by designing an environment merger with brittle install layout requirements, that fails BOTH at being what Nix tries to be, AND at being developer-friendly.

Yes, it is extra work for us but until we have a proper fix there is not much else we can do. Yes, it is potentially brittle and we have to deal with issues occasionally but it somehow works for the most part. There should not really be any extra work for upstreams as long as they follow the best practice of [not repeating](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) installation path constants. I should not have added the option but I opened https://github.com/GSConnect/gnome-shell-extension-gsconnect/pull/197 long time ago, when I have not yet had a clear concept of roles of upstream and downstream.

Reply:
> I should not have added the option but I opened https://github.com/GSConnect/gnome-shell-extension-gsconnect/pull/197 long time ago, when I have not yet had a clear concept of roles of upstream and downstream.

To be fair, neither did I :) Anyways, having a hacky script kicking around bothers me more than a meson switch, so I'm glad to get rid of it!

Reply:
Alright, so based on discussion it seems that the only real issue Nix has is that the un-optimized environment builder doesn't know how to exclude specific files, leading to a conflict when multiple store prefixes try to overlay the same file. The actual package itself does not care, and the software works if you don't create merged environments out of it.

So that's something which Nix already intends to fix, apparently, and when it's fixed then gnome-shell extensions on Nix can perfectly happily use `datadir/glib-2.0/schemas` too.

Whose responsibility it is to work around that can be debated, naturally.

For my part, I conclude that Meson probably does not need new functionality for `gnome.post_install()`.